src/data/qpf/univariate/pfunctor/M.lean

```diff
@@ -0,0 +1,794 @@
+by refl
+
+lemma iselect_cons [decidable_eq F.A] [inhabited (M F)] (ps : path F) {a} (f : F.B a → M F) {i} :
+  iselect (⟨a,i⟩ :: ps) (M.mk ⟨a,f⟩) = iselect ps (f i) :=
+by simp only [iselect,isubtree_cons]
+
+lemma corec_def {X} (f : X → F.obj X) (x₀ : X) :
+  M.corec f x₀ = M.mk (M.corec f <$> f x₀)  :=
+begin
+  dsimp [M.corec,M.mk],
+  congr, ext n,
+  cases n with n,
+  { dsimp [s_corec,approx.s_mk], refl, },
+  { dsimp [s_corec,approx.s_mk], cases h : (f x₀),
+    dsimp [s_corec._match_1,(<$>),pfunctor.map],
+    congr, }
+end
+
+lemma ext_aux [inhabited (M F)] [decidable_eq F.A] {n : ℕ} (x y z : M F)
+  (hx : agree' n z x)
+  (hy : agree' n z y)
+  (hrec : ∀ (ps : path F),
+             n = ps.length →
+            iselect ps x = iselect ps y) :
+  x.approx (n+1) = y.approx (n+1) :=
+begin
+  induction n with n generalizing x y z,
+  { specialize hrec [] rfl,
+    induction x using pfunctor.M.cases_on', induction y using pfunctor.M.cases_on',
+    simp only [iselect_nil] at hrec, subst hrec,
+    simp only [approx_mk, true_and, eq_self_iff_true, heq_iff_eq],
+    ext, },
+  { cases hx, cases hy,
+    induction x using pfunctor.M.cases_on', induction y using pfunctor.M.cases_on',
+    subst z,
+    replace hx_a_2 := mk_inj hx_a_2, cases hx_a_2,
+    replace hy_a_1 := mk_inj hy_a_1, cases hy_a_1,
+    replace hy_a_2 := mk_inj hy_a_2, cases hy_a_2,
+    simp [approx_mk], ext i, apply n_ih,
+    { apply hx_a_3 }, { apply hy_a_3 },
+    introv h, specialize hrec (⟨_,i⟩ :: ps) (congr_arg _ h),
+    simp [iselect_cons] at hrec, exact hrec }
+end
+
+open pfunctor.approx
+
+-- variables (F : pfunctor.{v})
+variables {F}
+
+local prefix `♯`:0 := cast (by simp [*] <|> cc <|> solve_by_elim)
+
+local attribute [instance, priority 0] classical.prop_decidable
+
+lemma ext [inhabited (M F)]
+  (x y : M F)
+  (H : ∀ (ps : path F), iselect ps x = iselect ps y) :
+  x = y :=
+begin
+  apply ext', intro i,
+  induction i with i,
+  { cases x.approx 0, cases y.approx 0, constructor },
+  { apply ext_aux x y x,
+    { rw ← agree_iff_agree', apply x.consistent },
+    { rw [← agree_iff_agree',i_ih], apply y.consistent },
+    introv H',
+    simp [iselect] at H,
+    cases H',
+    apply H ps }
+end
+
+section bisim
+  variable (R : M F → M F → Prop)
+  local infix ~ := R
+
+  /-- Bisimulation is the standard proof technique for equality between
+  infinite tree-like structures -/
+  structure is_bisimulation : Prop :=
+  (head : ∀ {a a'} {f f'}, M.mk ⟨a,f⟩ ~ M.mk ⟨a',f'⟩ → a = a')
+  (tail : ∀ {a} {f f' : F.B a → M F},
+    M.mk ⟨a,f⟩ ~ M.mk ⟨a,f'⟩ →
+    (∀ (i : F.B a), f i ~ f' i) )
+
+  theorem nth_of_bisim [inhabited (M F)] (bisim : is_bisimulation R) (s₁ s₂) (ps : path F) :
+       s₁ ~ s₂ →
+         is_path ps s₁ ∨ is_path ps s₂ →
+         iselect ps s₁ = iselect ps s₂ ∧
+         ∃ a (f f' : F.B a → M F),
+           isubtree ps s₁ = M.mk ⟨a,f⟩ ∧
+           isubtree ps s₂ = M.mk ⟨a,f'⟩ ∧
+         ∀ (i : F.B a), f i ~ f' i :=
```

Comment:
```suggestion
  theorem nth_of_bisim [inhabited (M F)] (bisim : is_bisimulation R) (s₁ s₂) (ps : path F)
       (h₀ : s₁ ~ s₂) (hh : is_path ps s₁ ∨ is_path ps s₂) :
     iselect ps s₁ = iselect ps s₂ ∧
     ∃ a (f f' : F.B a → M F), isubtree ps s₁ = M.mk ⟨a,f⟩ ∧ isubtree ps s₂ = M.mk ⟨a,f'⟩ ∧
       ∀ (i : F.B a), f i ~ f' i :=
```

Reply:
I'd rather leave this statement as is. I'd like to put `M.mk ⟨a,f⟩` above `M.mk ⟨a,f'⟩` because they are very subtly different and I'd like to emphasize that difference.

Reply:
:+1: